doc/neps/nep-0029-dtype-as-type.rst

```diff
@@ -0,0 +1,268 @@
+                return dtype_scalar_dict[obj]
+            elif len(args) < 1:
+                # Dtype('int8') or Dtype('S10') or record descr
+                return create_new_descr(cls, obj, *args, **kwargs)
+            else:
+                return super().__new__(cls, obj, *args, **kwargs)
+
+        def __call__(self, *args, **kwargs):
+            return self.typeobj(*args, **kwargs)
+
+    class IntDtype(Dtype):
+        def __repr__(self):
+            if self is IntDescr:
+                return type.__repr__(self)
+            return 'dtype(%s%d)' %(self.kind, self.itemsize)
+
+    class GenericDescr(type, metaclass=Dtype):
+        pass
+
+    class IntDescr(GenericDescr, metaclass=IntDtype):
+        def format(value):
+            return '%d' % value
+
+    class UInt8Descr(IntDescr):
+        kind = 'uint'
+        itemsize = 8
+        typeobj = np.uint8
+        # sort, fill, cast, clip, ...
+        ArrFuncs = int8_arrayfuncs
+
+    dtype_int_dict = {1: UInt8Descr}
+    dtype_scalar_dict = {np.uint8: UInt8Descr}
+
+At NumPy startup, as we do today, we would generate the builtin set of
+descriptor classes, and fill in ``dtype_int_dict`` and ``dtype_scalar_type``
+so that the built-in descriptors would continue to be singletons. ``Void``,
+``Byte`` and ``Unicode`` descriptors would be constructed on demand, as is done
+today.
+
+All dtype instances would inherit from ``GenericDescr`` which inherits from
+``type``, making them instances of ``type``::
+
+    >>> a = np.dtype(np.int8)
+    >>> a.mro(a)
+    [dtype(uint8), <class 'dtype.IntDescr'>, <class 'dtype.GenericDescr'>, \
+     <class 'type'>, <class 'object'>]
+    
+Each descr class will have its own set of ArrFuncs (``clip``, ``fill``,
+``cast``), The ``format`` function is what ``array_print`` will call to turn a
+memory location into a string.
+
+Downstream users of NumPy could subclass these type classes. Creating a categorical
+dtype would look like this (without error checking for out-of-bounds values)::
+
+    class Colors(UInt8Descr):
+        colors = ['red', 'green', 'blue']
+        def format(value):
+            return Colors.colors[value]
+        ArrFuncs = null_arrayfuncs
+
+    c = np.array([0, 1, 1, 0, 2], dtype=Colors)    
+
+Additional code would be needed to neutralize the `tp_as_number` slot functions.
+
+Advantages
+==========
+
+It is very difficult today to override dtype behaviour, since internally
+descriptor objects are not true type instances, rather contianers for the
+``ArrayDescrObject`` struct.
+
+Disadvantages
+=============
+
+Making descriptors into type objects requires thinking about type classes,
+which is more difficult to reason about than object instances. For instance,
+note that in the ``Colors`` example, we did not instantiate an object of the
+``Colors`` type, rather used that type directly in the ndarray creation. Also
+the ``format`` function is not a bound method of a class instance, rather an
```

Comment:
I think this can be circumvented of thinking of it as `__format__` and that what the print machinery would actually do (once we allow instances to be scalars) is `Color(array_scalar).__format__()`.

Reply:
The idea is that `arr.dtype.format` would replace the [`formatfunction` in arrayprint](https://github.com/numpy/numpy/blob/master/numpy/core/arrayprint.py#L497). I don't think we should be requiring scalar instantiation in each iteration of a loop to stringify array values.

This is a precursor to more extensively using method overloading to replace much of the case-based lookups scattered around the code, which is key to making subclasses of dtypes work.